src/modules/load_mon/LoadMon.cpp

```diff
@@ -0,0 +1,227 @@
+		stack_usage();
+	}
+
+	if (should_exit()) {
+		ScheduleClear();
+		exit_and_cleanup();
+	}
+
+	perf_end(_cycle_perf);
+}
+
+void LoadMon::cpuload()
+{
+	if (_last_idle_time == 0) {
+		// Just get the time in the first iteration */
+		_last_idle_time = system_load.tasks[0].total_runtime;
+		_last_idle_time_sample = hrt_absolute_time();
+		return;
+	}
+
+	irqstate_t irqstate = enter_critical_section();
+	const hrt_abstime now = hrt_absolute_time();
+	const hrt_abstime total_runtime = system_load.tasks[0].total_runtime;
+	leave_critical_section(irqstate);
+
+	// compute system load
+	const float interval = now - _last_idle_time_sample;
+	const float interval_idletime = total_runtime - _last_idle_time;
+
+	// get ram usage
+	struct mallinfo mem = mallinfo();
+	float ram_usage = (float)mem.uordblks / mem.arena;
+
+	cpuload_s cpuload{};
+	cpuload.load = 1.f - interval_idletime / interval;
+	cpuload.ram_usage = ram_usage;
+	cpuload.timestamp = hrt_absolute_time();
+
+	_cpuload_pub.publish(cpuload);
+
+	// store for next iteration
+	_last_idle_time = total_runtime;
+	_last_idle_time_sample = now;
+}
+
+void LoadMon::stack_usage()
+{
+	unsigned stack_free = 0;
+	unsigned fds_free = FDS_LOW_WARNING_THRESHOLD + 1;
+
+	bool checked_task = false;
+
+	task_stack_info_s task_stack_info{};
+	static_assert(sizeof(task_stack_info.task_name) == CONFIG_TASK_NAME_SIZE,
+		      "task_stack_info.task_name must match NuttX CONFIG_TASK_NAME_SIZE");
+
+	sched_lock();
+
+	if (system_load.tasks[_stack_task_index].valid && (system_load.tasks[_stack_task_index].tcb->pid > 0)) {
+
+		stack_free = up_check_tcbstack_remain(system_load.tasks[_stack_task_index].tcb);
+
+		strncpy((char *)task_stack_info.task_name, system_load.tasks[_stack_task_index].tcb->name, CONFIG_TASK_NAME_SIZE - 1);
+		task_stack_info.task_name[CONFIG_TASK_NAME_SIZE - 1] = '\0';
+
+		checked_task = true;
+
+#if CONFIG_NFILE_DESCRIPTORS > 0
+		FAR struct task_group_s *group = system_load.tasks[_stack_task_index].tcb->group;
+
+		unsigned tcb_num_used_fds = 0;
+
+		if (group) {
+			for (int fd_index = 0; fd_index < CONFIG_NFILE_DESCRIPTORS; ++fd_index) {
+				if (group->tg_filelist.fl_files[fd_index].f_inode) {
+					++tcb_num_used_fds;
+				}
+			}
+
+			fds_free = CONFIG_NFILE_DESCRIPTORS - tcb_num_used_fds;
+		}
+
+#endif // CONFIG_NFILE_DESCRIPTORS
+	}
+
+	sched_unlock();
+
+	if (checked_task) {
+		task_stack_info.stack_free = stack_free;
+		task_stack_info.timestamp = hrt_absolute_time();
+
+		_task_stack_info_pub.publish(task_stack_info);
+
+		// Found task low on stack, report and exit. Continue here in next cycle.
+		if (stack_free < STACK_LOW_WARNING_THRESHOLD) {
+			PX4_WARN("%s low on stack! (%i bytes left)", task_stack_info.task_name, stack_free);
+		}
+
+		// Found task low on file descriptors, report and exit. Continue here in next cycle.
+		if (fds_free < FDS_LOW_WARNING_THRESHOLD) {
+			PX4_WARN("%s low on FDs! (%i FDs left)", task_stack_info.task_name, fds_free);
```

Comment:
These warnings will now appear 3 times per second and spam the log quite a bit, right?

Reply:
> These warnings will now appear 3 times per second and spam the log quite a bit, right?

Only in the worst case if running the max number of tasks (32 or 64) and every one of them was either low on stack or FDs.

Overall it's kind of a lateral move. Currently in master it cycles at 1 Hz, but each iteration checks 2 tasks, and if a task entry (an array of CONFIG_MAX_TASKS) is empty it jumps forward. Now it runs slightly faster for the sake of cpuload measurement, but it's doing less work per iteration checking tasks.